ipld/src/selector/mod.rs

```diff
@@ -236,14 +244,116 @@ impl Selector {
     }
 
     /// Processes and returns resultant selector node
-    pub fn explore(self, _ipld: &Ipld, _p: &PathSegment) -> Option<Selector> {
-        // TODO
-        todo!()
+    pub fn explore(self, ipld: &Ipld, p: &PathSegment) -> Option<Selector> {
```

Comment:
This looks great, I only had a question on is selector a heavy enum? If the answer is yes what do you think about a reference?

It would also be interested(although abit premature right now) to see selector implement an iterator so that we could use it more lazily

Reply:
Not quite sure what you mean by `heavy enum`, care to elaborate? Yeah it would definitely be a more idiomatic interface to have this as an iterator, but I will hold off until the functionality is fully tested (in next PR). Great suggestion!

Reply:
Should have been more clear sorry! Something that occupies a lot of data, I guess what I am getting at is asking if the clones are going to be expensive but then again that is also a premature optimization

Reply:
Yeah, in practice it'll be very small (depends on the construction of the selector) but since it's recursive, there is no cap on the size. 

Unfortunately the clones are intended (as far as I can see) because of the certain branches which would modify selectors with requiring the previous to not be mutated. This is something I will explore very in depth to make sure there is not a more efficient way around this once I have a testsuite verifying all functionality.